File: lib/utils/single_instance.py
import os
import sys
import time
import psutil

from lib.utils.constants import lock_file
from lib.utils.singleton import singleton


@singleton
class SingleInstance:
    """Application single instance handler"""

    def acquire(self, retry=False) -> None:
        """Try to acquire lock file"""
        if os.path.exists(lock_file):
            with open(lock_file, "r") as f:
                pid = int(f.read().strip())

            if psutil.pid_exists(pid):
                if not retry:
                    time.sleep(1)
                    self.acquire(True)
                else:
                    print(f"Application already running with pid {pid}")
                    sys.exit(1)
            else:
                os.remove(lock_file)

        with open(lock_file, "w") as f:
            f.write(str(os.getpid()))


single_instance = SingleInstance()

File: RogPerfTuner/src/utils/single_instance.cpp
#include "../../include/utils/single_instance.hpp"

#include <signal.h>

#include <fstream>
#include <iostream>

#include "../../include/utils/constants.hpp"
#include "../../include/utils/file_utils.hpp"

void SingleInstance::acquire() {
	killRunningInstance();

	std::ofstream out(Constants::LOCK_FILE);
	out << Constants::PID;
	out.close();
}

bool SingleInstance::killRunningInstance() {
	auto running = false;
	if (FileUtils::exists(Constants::LOCK_FILE)) {
		std::ifstream f(Constants::LOCK_FILE);
		pid_t pid;
		f >> pid;
		f.close();

		if (pid > 0 && kill(pid, 0) == 0) {
			std::cout << "Found instance with pid " << pid << ", killing..." << std::endl;
			kill(pid, SIGKILL);
			running = true;
		}
		FileUtils::remove(Constants::LOCK_FILE);
	}
	return running;
}